### PR TITLE
add tcp, tcp-tls features + refactor of client, server

### PR DESCRIPTION
**Pull request by jkralik**

### CODE REVIEW

File: client.go [hunk 1]
```diff
@@ -0,0 +1,295 @@
...
+	case *net.TCPConn, *tls.Conn:
+		resp, _, err := PullTcp(m)
+		if err != nil {
+			return nil, err
+		}
+		return resp, nil
+	}
+	// UDP connection
+	resp, err := ParseDgramMessage(m)
+	if err != nil {
```
Comment (michalpristas) [inline client.go]: you can omit check and return resp, err

**Comment by michalpristas** [comment]
few pointers.
- read this: https://medium.com/@benbjohnson/standard-package-layout-7cdbc8391fc1
- try to use more granular files. they are long and not readable, example client.go, it contains definition of client, connection and some statics. crate client.go and conn.go 
- stick with interfaces, then create struct implementing interace. keep interfaces small. combine them to larger and more powerfull ones

this might be of use as well: https://www.youtube.com/watch?v=MzTcsI6tn-0

i can look into specifics later

File: README.md [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+		// for tcp
+		// co, err := coap.Dial("tcp", "localhost:5688")
+		
+		// for tcp-tls
+		// co, err := coap.DialWithTLS("localhost:5688", &tls.Config{InsecureSkipVerify: true})
+
+		if err != nil {
+			log.Fatalf("Error dialing: %v", err)
+		}
+		defer co.Close()
```
Comment (michalpristas) [inline README.md]: close the connection right away, defer for close might lead to running out of sockets.
also co.Close() error is ignored

File: client.go [hunk 3]
```diff
@@ -0,0 +1,295 @@
...
+			continue
+		} else if err != nil {
+			return nil, err
+		}
+		break
+	}
+	switch co.Conn.(type) {
+	case *net.TCPConn, *tls.Conn:
+		resp, _, err := PullTcp(m)
+		if err != nil {
```
Comment (michalpristas) [inline client.go]: same here return resp, err. turn on golint

File: client.go [hunk 4]
```diff
@@ -0,0 +1,295 @@
...
+func (c *Client) getTimeoutForRequest(timeout time.Duration) time.Duration {
+	var requestTimeout time.Duration
+	if c.Timeout != 0 {
+		requestTimeout = c.Timeout
+	} else {
+		requestTimeout = timeout
+	}
+	// net.Dialer.Timeout has priority if smaller than the timeouts computed so
+	// far
+	if c.Dialer != nil && c.Dialer.Timeout != 0 {
```
Comment (michalpristas) [inline client.go]: readability: single if?

File: client.go [hunk 5]
```diff
@@ -0,0 +1,295 @@
...
+		}
+	}
+	return requestTimeout
+}
+
+// Dial connects to the address on the named network.
+func Dial(network, address string) (conn *Conn, err error) {
+	client := Client{Net: network}
+	conn, err = client.Dial(address)
+	if err != nil {
```
Comment (michalpristas) [inline client.go]: return conn, err

Comment (michalpristas) [reply client.go]: even better, return client.DIal()

File: client.go [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+		return nil, err
+	}
+	return conn, nil
+}
+
+// DialTimeout acts like Dial but takes a timeout.
+func DialTimeout(network, address string, timeout time.Duration) (conn *Conn, err error) {
+	client := Client{Net: network, Dialer: &net.Dialer{Timeout: timeout}}
+	conn, err = client.Dial(address)
```
Comment (michalpristas) [inline client.go]: same

File: client.go [hunk 7]
```diff
@@ -0,0 +1,295 @@
...
+	return conn, nil
+}
+
+// DialWithTLS connects to the address on the named network with TLS.
+func DialWithTLS(network, address string, tlsConfig *tls.Config) (conn *Conn, err error) {
+	if !strings.HasSuffix(network, "-tls") {
+		network += "-tls"
+	}
+	client := Client{Net: network, TLSConfig: tlsConfig}
+	conn, err = client.Dial(address)
```
Comment (michalpristas) [inline client.go]: client.Dial(address)

File: client.go [hunk 8]
```diff
@@ -0,0 +1,295 @@
...
+	return conn, nil
+}
+
+// DialTimeoutWithTLS acts like DialWithTLS but takes a timeout.
+func DialTimeoutWithTLS(network, address string, tlsConfig *tls.Config, timeout time.Duration) (conn *Conn, err error) {
+	if !strings.HasSuffix(network, "-tls") {
+		network += "-tls"
+	}
+	client := Client{Net: network, Dialer: &net.Dialer{Timeout: timeout}, TLSConfig: tlsConfig}
+	conn, err = client.Dial(address)
```
Comment (michalpristas) [inline client.go]: client.Dial(address)

File: message.go [hunk 9]
```diff
@@ -0,0 +1,670 @@
...
+	case int:
+		v = uint32(i)
+	case int32:
+		v = uint32(i)
+	case uint:
+		v = uint32(i)
+	case uint32:
+		v = i
+	default:
+		panic(fmt.Errorf("invalid type for option %x: %T (%v)",
```
Comment (michalpristas) [inline message.go]: if this is not really needed consider returning error instead of panicing

File: messagetcp.go [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+	if err := binary.Read(r, binary.BigEndian, &firstByte); err != nil {
+		return mti, normalizeErrors(err)
+	}
+	hdrOff++
+
+	lenNib := (firstByte & 0xf0) >> 4
+	tkl := firstByte & 0x0f
+
+	var opLen int
+	if lenNib < TCP_MESSAGE_LEN13_BASE {
```
Comment (michalpristas) [inline messagetcp.go]: readability: switch?

File: server.go [hunk 11]
```diff
@@ -0,0 +1,720 @@
...
+		}
+	}
+
+	defer atomic.AddInt32(&srv.workersCount, -1)
+
+	inUse := false
+	timeout := time.NewTimer(idleWorkerTimeout)
+	defer timeout.Stop()
+LOOP:
+	for {
```
Comment (michalpristas) [inline server.go]: try to avoid this, maybe extract to separate func and return instead of break ANCH

File: server.go [hunk 12]
```diff
@@ -0,0 +1,720 @@
...
+		// Check PacketConn interface's type is valid and value
+		// is not nil
+		if t, ok := pConn.(*net.UDPConn); ok && t != nil {
+			if e := setUDPSocketOptions(t); e != nil {
+				return e
+			}
+			srv.started = true
+			srv.lock.Unlock()
+			e := srv.serveUDP(t)
+			srv.lock.Lock() // to satisfy the defer at the top
```
Comment (michalpristas) [inline server.go]: try to think about this seems confusing

File: client.go (hunk 1 above)
Comment (jkralik) [reply client.go]: done

File: README.md (hunk 2 above)
Comment (jkralik) [reply README.md]: Yeah I understand. For example i removed this line because it is simple example

File: client.go (hunk 3 above)
Comment (jkralik) [reply client.go]: Done

File: client.go (hunk 4 above)
Comment (jkralik) [reply client.go]: Done

File: client.go (hunk 5 above)
Comment (jkralik) [reply client.go]: Done

File: client.go (hunk 6 above)
Comment (jkralik) [reply client.go]: Done

File: client.go (hunk 7 above)
Comment (jkralik) [reply client.go]: Done

File: client.go (hunk 8 above)
Comment (jkralik) [reply client.go]: Done

File: message.go (hunk 9 above)
Comment (jkralik) [reply message.go]: Now it returns error and function upper just log fatal error for that option.

File: messagetcp.go (hunk 10 above)
Comment (jkralik) [reply messagetcp.go]: done

File: server.go (hunk 11 above)
Comment (jkralik) [reply server.go]: done

File: server.go (hunk 12 above)
Comment (jkralik) [reply server.go]: done

**Pull request merged**